bpo-26806: add 30 to the recursion limit in IDLE's shell

taleinat opened the pull request:
This is done to compensate for the extra stack frames added by IDLE's shell, which can cause problems when setting the recursion limit to low values.

This wraps `sys.setrecursionlimit()` and `sys.getrecursionlimit()`, adding a note about this to their doc-strings.

<!-- issue-number: [bpo-26806](https://bugs.python.org/issue26806) -->
https://bugs.python.org/issue26806
<!-- /issue-number -->

taleinat committed:
bpo-26806: transparently add 25 to the recursion limit in IDLE

This is done to compensate for the extra stack frames added by
IDLE itself, which cause problems when setting the recursion limit
to low values.

This wraps sys.setrecursionlimit() and sys.getrecursionlimit()
as invisibly as possible.

taleinat committed:
add a test + bugfix

taleinat committed:
add docstring + comment referencing the bpo issue

terryjreedy commented on Lib/idlelib/run.py:
```diff
@@ -474,6 +475,44 @@ def __init__(self, rpchandler):
+            del sys
+            del wraps
+            """)
+        exec(code, self.locals)
```
I believe that this code should be executed in the run module on import, NOT in the main module.  Calling a module function from main, as I suggested on the issue, should do that.  Then the deletions are not needed.

terryjreedy replied:
Done, except that call to install_recursionlimit_wrappers is in MyHandler.handle rather than main().  I presume the rationale is that 4 other sys attributes are replaced here.  Good enough.

terryjreedy reviewed (changes requested):
Also see new comment on the issue, which includes some manual tests.

taleinat committed:
address review comments

1. define wrapping code at the module level
2. increase recursion limit delta from 25 to 30
3. add a note to the wrapped functions' doc-strings
4. extract the recursion limit delta to a variable
5. rework test and add another one

taleinat committed:
minor fix to uninstall_recursionlimit_wrappers()

taleinat committed:
update NEWS

terryjreedy commented on Lib/idlelib/run.py:
```diff
@@ -303,6 +305,60 @@ def fix_scaling(root):
+            limit, = args
+        except ValueError:
+            raise TypeError("setrecursionlimit() takes exactly one "
+                            "argument ({} given)".format(len(args)))
```
Use f-string here also. 
<DONE, and added bad call tests>

Messages verified.  It is somewhat annoying that C and python functions with same signature (here, 'limit, /') give different error messages for the same error.  (A separate issue for the tracker.)  So we cannot, even for 3.8+, when '/' is valid, match behavior by just giving the proper signature.

terryjreedy commented on Lib/idlelib/run.py:
```diff
@@ -303,6 +305,60 @@ def fix_scaling(root):
+
+
+def uninstall_recursionlimit_wrappers():
+    """Uninstall the recursion limit wrappers from the sys module."""
```
This is (only) needed for tests.  Add line "# Needed for tests." so function is never deleted as 'unused'.  (I have made similar notes for other test-only code.)
<Done.>

taleinat replied:
Why would it be deleted if it is used in test code? Even if it were deleted, the tests would immediately break!

Also, there's no telling who might find other uses for this in the future.

terryjreedy replied:
Let me start over.  I initially wasted time looking for where this was used in run.py.   I like to mark test code (code we/IDLE only use as such).  So I moved the comment, reworded, into the docstring *and* added a note that users could also use this.  (Yeah, one of use might want to do so sometime.)  I think that this is an improvement.

terryjreedy commented on Misc/NEWS.d/next/IDLE/2019-06-10-22-48-50.bpo-26806.Zltkum.rst:
```diff
@@ -0,0 +1,2 @@
+Always add 30 to the recursion limit in IDLE's shell to compensate for
+stack frames added by IDLE.
```
Suggestion:
"To compensate for stack frames added by IDLE and avoid possible problems with low recursion limits, add 30 to limits in the user code execution process.  Subtract 30 when reporting recursion limits to make this addition mostly transparent."
<Done.>
Note: we are not doing the same in the IDLE process when -n (deprecated) is passed.

terryjreedy committed:
Merge remote-tracking branch 'origin/master' into pr_13944

terryjreedy committed:
Use f-string for error message and test bad calls.

terryjreedy committed:
Add test note.

terryjreedy reviewed (commented):
When I initially pulled this PR and ran test_idle, I got 4 failures.  They disappeared when I merged in current master.  We need to remember to do this before final testing.

I have made the minor changes I suggested and added a test method.  I will add the idlelib news item and probably a doc change and then push everything for your final approval.

terryjreedy committed:
Revise blurb and add idlelib NEWS item.

terryjreedy committed:
Add paragraph to doc.

terryjreedy reviewed (approved):
Please merge if changes seem OK.

taleinat commented on Doc/library/idle.rst:
```diff
@@ -713,6 +713,10 @@ or ``print`` or ``write`` to sys.stdout or sys.stderr,
 
+The IDLE code running in the execution process adds frames to the call stack
+that would not be there otherwise.  IDLE wraps ``sys.getrecursionlimit`` and
+``sys.setrecursionlimit`` to reduce their visibility.
```
"their" is ambiguous here; I suggest "to reduce the visibility of the additional stack frames."

Also, perhaps we should mention that this is not just about visibility? E.g. "to reduce the effect of the additional stack frames".

terryjreedy committed:
Update idle.rst

terryjreedy committed:
Update run.py

taleinat committed:
fix typo

taleinat committed:
fix whitespace

taleinat committed:
another whitespace fix
